src/engine/server_iv.c

```diff
@@ -1128,8 +1128,8 @@ iv_op(struct ds_iv_ns *ns, struct ds_iv_key *key, d_sg_list_t *value,
 		 */
 		D_WARN("retry upon %d for class %d opc %d\n", rc,
 		       key->class_id, opc);
-		/* Yield to avoid hijack the cycle if IV RPC is not sent */
-		ABT_thread_yield();
+		/* sleep 1sec and retry */
+		dss_sleep(1000);
```

Comment:
It just reduced some retries, I don't see much difference with original yield.

The problem is that in DAOS-9040 we know the IV retries were not from this iv_op() function, since we didn't see above warning messages in the log, any thoughts?

Reply:
oh, you need check the log on the PS leader, where are those retry log generated. (wolf-111)  

src/engine/server_iv.c:1049 sync_comp_cb() retry for class 4 opc 2 rc DER_GRPVER(-1036): 'Group versioning mismatch'

If we use yield(), it will keep retry anyway.

Reply:
then I'm wondering how can this yield or sleep help? the retry happens in sync_comp_cb(), which is before above iv_op_internal() returns, so this yield or sleep will never be called, right? (and I didn't see the warning messages before this yield call)

Reply:
ah, just updated the patch, please check. Thanks.

Reply:
The patch https://github.com/daos-stack/daos/pull/7473 is in testing, I think it is more direct and simple.

Reply:
@Nasf-Fan  we need retry for asynchronous case.  So sleep 1 second, then retry will reduce the GRPVER message. But again this patch will only reduce the GRPVER message, and I still do not understand why pool map IV sync is slow.

Reply:
Some confused, in the patch #7473, it still retry for asynchronous case with yield, anything wrong?

Reply:
I agree with @Nasf-Fan that this part of retry code is confusing: Given It has been retried in sync_comp_cb() for DER_GRPVER, I think it's not necessary to retry (here for the same error), right?

The other thing I don't quite follow is: there is already an IV completion callback (ds_iv_done()) called on IV operation completion, why do we need another callback (sync_comp_cb()) which is also called on IV operation completion? The ds_iv_done() is used to wake up caller (in sync mode), but the the sync_comp_cb() is called in a  unnamed ULT, which could be scheduled at a random time after the IV operation completed?

It looks to me this sync_comp_cb() should be merged into the IV op completion callback to make code clear and avoid potential race.

Reply:
Sorry, not sure I follow the question here.   So we have lazy mode and eager mode for IV operation.

For lazy mode,  iv_op_internal() will not execute the IV sync operation, and sync_comp_cb() is the function to be called when the asynchronous IV finish,   so we will retry here.

For eager mode, it will retry inside iv_op_internal()  anyway.

Sync_comp_cb() is initially implemented for releasing the resource for asynchronous IV sync.  ds_iv_done is the eventual completion callback for iv_op itself.  Not sure how to merge them together, hmm

Reply:
>I agree with @Nasf-Fan that this part of retry code is confusing: Given It has been retried in sync_comp_cb() for DER_GRPVER, I think it's not necessary to retry (here for the same error), right?

Those are for totally different cases, and we have to retry in both cases.  or I miss your question?

Reply:
The condition for retry in sync_comp_cb() is "if (cb_arg->retry && rc == -DER_GRPVER && !cb_arg->ns->iv_stop)".

The condition for retry in iv_op() is "if (retry && !ns->iv_stop && (daos_rpc_retryable_rc(rc) || rc == -DER_NOTLEADER))".

The former one is the subset of the latter one. As long as the condition in sync_comp_cb() can be triggered, then it also matches the retry condition in iv_op(). Then why not directly use the retry logic inside iv_op()?

Reply:
>The former one is the subset of the latter one. As long as the condition in sync_comp_cb() can be triggered, then it also matches the retry condition in iv_op(). Then why not directly use the retry logic inside iv_op()?

Hmm, sorry, I do not see big difference here. Just two if check?